### PR TITLE
Tabular: Added out-of-fold methods to predictor.py

### PR DESCRIPTION
**Pull request by Innixma**
*Issue #, if available:*

*Description of changes:*

- Added out-of-fold methods to Predictor
- Added positive_class() method to Predictor (curious on if it should be property, variable, or method)
- Minor PEP8 cleanup

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by szha** [comment]
Job PR-779-4 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-779/4/index.html

File: tabular/src/autogluon/tabular/task/tabular_prediction/predictor.py [hunk 1]
```diff
@@ -816,6 +817,120 @@ def fit_weighted_ensemble(self, base_models: list = None, name_suffix='_custom',
...
+                if not transformed:
+                    y_pred_proba_oof_transformed.columns = copy.deepcopy(self._learner.label_cleaner.ordered_class_labels)
+        if transformed:
+            return y_pred_proba_oof_transformed
+        else:
+            return self.transform_labels(labels=y_pred_proba_oof_transformed, inverse=True, proba=True)
+
+    # TODO: v0.1 Should this instead return None when not valid instead of raising exception? Should this be a property or variable instead?
+    # TODO: v0.1 Add positive_class parameter to task.fit
+    def positive_class(self):
```
Comment (jwmueller) [inline tabular/src/autogluon/tabular/task/tabular_prediction/predictor.py:921]: Tutorials should be updated to reflect this new way of getting positive class.

File: tabular/src/autogluon/tabular/task/tabular_prediction/predictor.py [hunk 2]
```diff
@@ -816,6 +817,120 @@ def fit_weighted_ensemble(self, base_models: list = None, name_suffix='_custom',
...
+            if as_multiclass and self.problem_type == BINARY:
+                y_pred_proba_oof_transformed = LabelCleanerMulticlassToBinary.convert_binary_proba_to_multiclass_proba(y_pred_proba_oof_transformed, as_pandas=True)
+                if not transformed:
+                    y_pred_proba_oof_transformed.columns = copy.deepcopy(self._learner.label_cleaner.ordered_class_labels)
+        if transformed:
+            return y_pred_proba_oof_transformed
+        else:
+            return self.transform_labels(labels=y_pred_proba_oof_transformed, inverse=True, proba=True)
+
+    # TODO: v0.1 Should this instead return None when not valid instead of raising exception? Should this be a property or variable instead?
```
Comment (jwmueller) [inline tabular/src/autogluon/tabular/task/tabular_prediction/predictor.py:919]: I prefer property, but don't have strong opinions either way. I think it should at least print warning if it's going to return None.

File: tabular/src/autogluon/tabular/task/tabular_prediction/predictor.py [hunk 3]
```diff
@@ -816,6 +817,120 @@ def fit_weighted_ensemble(self, base_models: list = None, name_suffix='_custom',
...
+            A more benign version of this issue occurs when 'log_loss' wasn't specified as the eval_metric but rare classes were dropped by AutoGluon.
+            In this case, not all of the original training data rows will have an OOF prediction. It is recommended to either drop these rows during the join or to get direct predictions on the missing rows via `predictor.predict`.
+
+        Parameters
+        ----------
+        model : str (optional)
+            The name of the model to get out-of-fold predictions from. Defaults to None, which uses the highest scoring model on the validation set.
+            Valid models are listed in this `predictor` by calling `predictor.get_model_names()`
+        transformed : bool, default = False
+            Whether the output values should be of the external label representation (False) or the internal label representation (True).
```
Comment (jwmueller) [inline tabular/src/autogluon/tabular/task/tabular_prediction/predictor.py:845]: Clarify external label representation = original class names, internal label representation = integers numbering the possible classes from 1,...,K

File: tabular/src/autogluon/tabular/task/tabular_prediction/predictor.py [hunk 4]
```diff
@@ -816,6 +817,120 @@ def fit_weighted_ensemble(self, base_models: list = None, name_suffix='_custom',
...
 
+    # TODO: v0.1 This can cause very strange issues related to index mismatches with original training data on extreme edge cases (multiclass where autogluon duplicated rows in data due to rare classes and eval_metric was log_loss)
+    #  This is a very rare case but needs to be fixed by v0.1 release
+    #  It might be good enough to do an inner join on training data, but it needs to be tested
+    def get_oof_pred(self, model: str = None, transformed=False) -> pd.Series:
+        """
+        Note: This is advanced functionality not intended for normal usage.
+
+        Returns the out-of-fold (OOF) training predictions of a model.
+        OOF predictions are unbiased estimates of model predictions on the training data.
```
Comment (jwmueller) [inline tabular/src/autogluon/tabular/task/tabular_prediction/predictor.py:828]: I think these sentences could be more clearly written as:
"Returns the out-of-fold (OOF) predictions for every row in the training data. OOF predictions may provide unbiased estimates of generalization accuracy (reflecting how predictions will behave on new data), as predictions for each row are only made using models that were fit to subset of data where this row was held-out"

File: tabular/src/autogluon/tabular/task/tabular_prediction/predictor.py [hunk 5]
```diff
@@ -816,6 +817,120 @@ def fit_weighted_ensemble(self, base_models: list = None, name_suffix='_custom',
...
+            return self.transform_labels(labels=y_pred_oof_transformed, inverse=True, proba=False)
+
+    # TODO: Improve error messages when trying to get oof from refit_full and distilled models.
+    # TODO: v0.1 add tutorial related to this method, as it is very powerful.
+    # TODO: v0.1 This can cause very strange issues related to index mismatches with original training data on extreme edge cases (multiclass where autogluon duplicated rows in data due to rare classes and eval_metric was logloss)
+    #  This is a very rare case but needs to be fixed by v0.1 release
+    #  It might be good enough to do an inner join on training data, but it needs to be tested
+    def get_oof_pred_proba(self, model: str = None, transformed=False, as_multiclass=False) -> Union[pd.DataFrame, pd.Series]:
+        """
+        Note: This is advanced functionality not intended for normal usage.
```
Comment (jwmueller) [inline tabular/src/autogluon/tabular/task/tabular_prediction/predictor.py:868]: this method's documentation can probably be much shorter and just say to refer to the `get_oof_pred()` documentation

File: tabular/src/autogluon/tabular/task/tabular_prediction/predictor.py (hunk 2 above)
Comment (Innixma) [reply tabular/src/autogluon/tabular/task/tabular_prediction/predictor.py:919]: Added

File: tabular/src/autogluon/tabular/task/tabular_prediction/predictor.py (hunk 5 above)
Comment (Innixma) [reply tabular/src/autogluon/tabular/task/tabular_prediction/predictor.py:868]: Added

File: tabular/src/autogluon/tabular/task/tabular_prediction/predictor.py (hunk 4 above)
Comment (Innixma) [reply tabular/src/autogluon/tabular/task/tabular_prediction/predictor.py:828]: Added

File: tabular/src/autogluon/tabular/task/tabular_prediction/predictor.py (hunk 1 above)
Comment (Innixma) [reply tabular/src/autogluon/tabular/task/tabular_prediction/predictor.py:921]: Good catch, added. Let me know if it needs to be changed in any other tutorials.

File: tabular/src/autogluon/tabular/task/tabular_prediction/predictor.py (hunk 3 above)
Comment (Innixma) [reply tabular/src/autogluon/tabular/task/tabular_prediction/predictor.py:845]: Added

**Comment by szha** [comment]
Job PR-779-5 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-779/5/index.html

**Comment by szha** [comment]
Job PR-779-7 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-779/7/index.html

File: tabular/src/autogluon/tabular/task/tabular_prediction/predictor.py [hunk 6]
```diff
@@ -816,6 +817,114 @@ def fit_weighted_ensemble(self, base_models: list = None, name_suffix='_custom',
...
+
+        Returns the out-of-fold (OOF) prediction probabilities for every row in the training data.
```
Comment (jwmueller) [inline tabular/src/autogluon/tabular/task/tabular_prediction/predictor.py:860]: would change "prediction probabilities" -> "predicted class probabilities"

File: tabular/src/autogluon/tabular/task/tabular_prediction/predictor.py (hunk 1 above)
Comment (jwmueller) [reply tabular/src/autogluon/tabular/task/tabular_prediction/predictor.py:914]: Still missing update in the quick-start tutorial as well. Search for the cell:
```
pred_probs = predictor.predict_proba(test_data_nolab)
positive_class = [label for label in predictor.class_labels if predictor.class_labels_internal_map[label]==1][0]  # which label is considered 'positive' class
print(f"Predicted probabilities of class '{positive_class}':", pred_probs)
```

File: tabular/src/autogluon/tabular/task/tabular_prediction/predictor.py [hunk 7]
```diff
@@ -816,6 +817,114 @@ def fit_weighted_ensemble(self, base_models: list = None, name_suffix='_custom',
...
+            return y_pred_proba_oof_transformed
+        else:
+            return self.transform_labels(labels=y_pred_proba_oof_transformed, inverse=True, proba=True)
+
+    # TODO: v0.1 Properly error/return None if label_cleaner hasn't been fit yet. (After API refactor)
+    # TODO: v0.1 Add positive_class parameter to task.fit
+    @property
+    def positive_class(self):
+        """
+        Returns the positive class name in binary classification. Useful for computing metrics such as F1 which require a positive and negative class.
```
Comment (jwmueller) [inline tabular/src/autogluon/tabular/task/tabular_prediction/predictor.py:916]: also clarify: "In binary classification, `predict_proba()` returns the estimated probability that each row belongs to the positive class.

Comment (Innixma) [reply tabular/src/autogluon/tabular/task/tabular_prediction/predictor.py:916]: added

File: tabular/src/autogluon/tabular/task/tabular_prediction/predictor.py (hunk 6 above)
Comment (Innixma) [reply tabular/src/autogluon/tabular/task/tabular_prediction/predictor.py:860]: added

File: tabular/src/autogluon/tabular/task/tabular_prediction/predictor.py (hunk 1 above)
Comment (Innixma) [reply tabular/src/autogluon/tabular/task/tabular_prediction/predictor.py:914]: added

**Review by jwmueller — APPROVED** [review]
looks good

**Comment by szha** [comment]
Job PR-779-8 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-779/8/index.html

**Pull request merged**